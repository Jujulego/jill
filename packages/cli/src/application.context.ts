import { createContext, FC, useContext } from 'react';
import yargs from 'yargs';

// Test
export type Args<A> = yargs.ArgumentsCamelCase<A> & { '--': readonly (string | number)[] };
export type Builder<A> = (yargs: yargs.Argv) => yargs.Argv<A>;
export type BuilderWrapper<A> = <Args>(yargs: yargs.Argv<Args>) => yargs.Argv<Omit<Args, keyof A> & A>;

export interface GlobalArgs {
  plugins: string[];
  verbose: number;
}

export interface Command<A> {
  id: string;
  name: string | readonly [string, ...string[]];
  description: string;
  builder: Builder<A>;
}

export interface ApplicationContextState {
  args: Args<unknown>;
  command?: Command<unknown>;
}

export type CommandComponent<A, P> = FC<P> & {
  command: Command<A>;
}

export type UseArgsHook<A> = () => Args<A & GlobalArgs>;

// Context
export const applicationDefaultState: ApplicationContextState = {
  args: {
    '$0': '',
    '_': [],
    '--': []
  }
};

export const ApplicationContext = createContext(applicationDefaultState);

// Hooks
export function useArgs<A>(): Args<A & GlobalArgs> {
  const { args } = useContext(ApplicationContext);
  return args as Args<A & GlobalArgs>;
}
