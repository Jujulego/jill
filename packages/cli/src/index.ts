export { buildCommand } from './commands/build';
export { eachCommand } from './commands/each';
export { infoCommand } from './commands/info';
export { logger, OraLogger } from './logger';
