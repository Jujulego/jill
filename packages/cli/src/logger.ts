import { logger as coreLogger } from '@jujulego/jill-core';
import { format, Logger } from 'winston';
import Transport from 'winston-transport';
import ora from 'ora';
import chalk from 'chalk';

// Constants
const MESSAGE = Symbol.for('message');

// Transport
export class OraTransport extends Transport {
  // Attributes
  readonly _spinner = ora();

  // Methods
  private _keepSpinner<T>(fun: () => T): T {
    // Save state
    let spinning = false;
    let text = '';

    if (this._spinner.isSpinning) {
      spinning = true;
      text = this._spinner.text;
    }

    try {
      return fun();
    } finally {
      // Restore state
      if (!this._spinner.isSpinning && spinning) {
        this._spinner.start(text);
      }
    }
  }

  log(info: any, next: () => void): void {
    // Print message
    const msg = info[MESSAGE] as string;

    this._keepSpinner(() => {
      this._spinner.stop();

      for (const line of msg.split('\n')) {
        process.stderr.write(line + '\n');
      }
    });

    next();
  }

  spin(message: string): void {
    this._spinner.start(message);
  }

  succeed(log: string): void {
    this._spinner.succeed(log);
  }

  fail(log: Error | string): void {
    if (typeof log === 'string') {
      this._spinner.fail(log);
    } else {
      this._spinner.fail(log.stack || log.toString());
    }
  }

  stop(): void {
    this._spinner.stop();
  }
}

// Logger
export class OraLogger {
  // Logger
  constructor(
    private readonly logger: Logger,
    private readonly transport: OraTransport
  ) {}

  // Methods
  // - logger
  debug(message: string): void {
    this.logger.debug({ message });
  }

  verbose(message: string): void {
    this.logger.verbose({ message });
  }

  info(message: string): void {
    this.logger.info({ message });
  }

  warn(message: string): void {
    this.logger.warn({ message });
  }

  error(message: string): void {
    this.logger.error({ message });
  }

  child(options: Record<string, unknown>): Logger {
    return this.logger.child(options);
  }

  // - ora
  spin(msg: string): void {
    this.transport.spin(msg);
  }

  succeed(msg: string): void {
    this.transport.succeed(msg);
  }

  fail(msg: string): void {
    this.transport.fail(msg);
  }

  stop(): void {
    this.transport.stop();
  }

  // Properties
  get level(): string {
    return this.logger.level;
  }

  set level(level: string) {
    this.logger.level = level;
  }
}

// Setup
export const transport = new OraTransport({
  format: format.combine(
    format.colorize({ message: true, colors: { debug: 'grey', verbose: 'blue', info: 'white', error: 'red' } }),
    format.printf(({ label, message }) => message.split('\n').map(line => [label && chalk.grey(`[${label}]`), line].filter(p => p).join(' ')).join('\n')),
  )
});

coreLogger.add(transport);

export const logger = new OraLogger(coreLogger, transport);
