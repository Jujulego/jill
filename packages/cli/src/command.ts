import { logger } from '@jujulego/jill-core';
import yargs from 'yargs';

import { transport } from './logger';

// Export
export type Awaitable<T> = T | PromiseLike<T>;
export type Arguments<A> = yargs.Arguments<A> & { '--': readonly (string | number)[] };
export type Builder<T, A = T> = (yargs: yargs.Argv<T>) => yargs.Argv<A>;

// Command
export abstract class Command<A = unknown> {
  // Attributes
  readonly logger = logger;
  readonly spinner = transport.spinner;

  // Methods
  protected abstract define<T, U>(builder: Builder<T, U>): Builder<T, U & A>;
  protected abstract run(args: Arguments<A>): Awaitable<number | void>;

  setup<T>(yargs: yargs.Argv<T>): yargs.Argv<T> {
    yargs.command<A>(
      this.name,
      this.description,
      (y) => this.define(_ => _)(y),
      (a) => this._wrapper(a as Arguments<A>)
    );

    return yargs;
  }

  private async _wrapper(args: Arguments<A>): Promise<void> {
    try {
      const exit = await this.run(args);
      process.exit(exit ?? 0);
    } catch (err) {
      this.spinner.fail(err.message);
      process.exit(1);
    }
  }

  log(msg: string): void {
    if (this.spinner.isSpinning) {
      this.spinner.clear();
    }

    console.log(msg);
  }

  // Properties
  abstract get name(): string | readonly string[];
  abstract get description(): string;
}