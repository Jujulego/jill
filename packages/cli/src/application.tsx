import { Children, FC, isValidElement, ReactElement, useEffect, useState } from 'react';
import Spinner from 'ink-spinner';
import { Text } from 'ink';
import yargs from 'yargs';
import { hideBin } from 'yargs/helpers';

import { CommandComponent } from './command.hoc';
import { ApplicationContext, ApplicationContextState, applicationDefaultState, Args } from './application.context';

// Types
export interface ApplicationProps {
  name: string;
}

// Constants
const parser = yargs(hideBin(process.argv))
  .parserConfiguration({
    'populate--': true,
  });

const commands = new Map<string, ReactElement>();

// Component
export const Application: FC<ApplicationProps> = ({ name, children }) => {
  // State
  const [state, setState] = useState<ApplicationContextState>(applicationDefaultState);

  // Effects
  useEffect(() => {
    // Config yargs
    parser.scriptName(name);

    // Define commands
    Children.map(children, (child) => {
      if (!isValidElement(child)) {
        return;
      }

      const { command } = child.type as CommandComponent<unknown, unknown>;
      commands.set(command.id, child);
      parser.command(
        command.name,
        command.description,
        (y) => command.builder(y),
        (args) => {
          setState((old) => ({
            ...old,
            args: args as Args<unknown>,
            command
          }));
        }
      );
    });

    // Parse to run command
    parser.strictCommands()
      .help()
      .parse();
  }, [name, children]);

  // Render
  if (!state.command) {
    return (
      <Text>
        <Spinner />
        {' '} Loading { name } ...
      </Text>
    );
  }

  return (
    <ApplicationContext.Provider value={state}>
      { commands.get(state.command.id) }
    </ApplicationContext.Provider>
  );
};
