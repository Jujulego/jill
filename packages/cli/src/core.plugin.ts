import { Plugin } from '@jujulego/jill-common';

import { EachCommand } from './commands/each.command';
import { RunCommand } from './commands/run.command';

// Plugin
export const corePlugin = Plugin.createPlugin('core', [
  RunCommand,
  EachCommand,
]);
