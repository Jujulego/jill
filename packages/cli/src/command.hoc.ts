import { FC } from 'react';

import { Args, Command, useArgs } from './application.context';

// Types
export type CommandMetadata<A> = Omit<Command<A>, 'id'>

export type CommandComponent<A, P> = FC<P> & {
  command: Command<A>;
}

export interface CommandUtils<A> {
  useArgs: () => Args<A>;
  wrapper: <P>(Component: FC<P>) => CommandComponent<A, P>;
}

// Generate HOC & Hook for command components
export function command<A>(command: CommandMetadata<A>): CommandUtils<A> {
  return {
    useArgs: () => useArgs<A>(),
    wrapper: <P>(Component: FC<P>): CommandComponent<A, P> => {
      Component.displayName = `command(${Component.displayName || Component.name})`;

      return Object.assign(Component, {
        command: {
          ...command,
          id: typeof command.name === 'string' ? command.name : command.name[0],
        }
      });
    }
  };
}
