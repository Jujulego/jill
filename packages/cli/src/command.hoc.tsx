import { FC, useContext } from 'react';
import yargs from 'yargs';
import { ArgsContext } from './application';

// Types
export type Arguments<A> = yargs.Arguments<A> & { '--': readonly (string | number)[] };
export type Builder<A> = (yargs: yargs.Argv) => yargs.Argv<A>;

export interface CommandMetadata<A> {
  name: string | readonly [string, ...string[]];
  description: string;
  define: Builder<A>;
}

export type CommandComponent<A, P = Record<string, unknown>> = FC<P> & {
  command: CommandMetadata<A> & { id: string };
}

// HOC
export function command<A, P = Record<string, unknown>>(command: CommandMetadata<A>, Component: FC<P & Arguments<A>>): CommandComponent<A, P> {
  const Wrapper: FC<P> = (props) => {
    const args = useContext(ArgsContext) as Arguments<A>;

    return <Component {...props} {...args} />;
  };

  Wrapper.displayName = `command(${Component.name || Component.displayName})`;

  return Object.assign(Wrapper, {
    command: {
      ...command,
      id: typeof command.name === 'string' ? command.name : command.name[0]
    }
  });
}
