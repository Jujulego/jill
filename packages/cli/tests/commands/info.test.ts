import { Project, Workspace } from '@jujulego/jill-core';
import chalk from 'chalk';
import yargs from 'yargs';

import { InfoCommand } from '../../src';
import { transport } from '../../src/logger';

// Setup
jest.mock('../../src/logger');

chalk.level = 1;

let cmd: InfoCommand;
let project: Project;
let screen: string;

beforeEach(() => {
  cmd = new InfoCommand(yargs);
  project = new Project('.');
  screen = '';

  // Mocks
  jest.resetAllMocks();
  jest.restoreAllMocks();

  jest.spyOn(console, 'log').mockImplementation((message) => screen += message + '\n');
  jest.spyOn(cmd, 'project', 'get').mockReturnValue(project);
});

// Tests
describe('jill info', () => {
  it('should exit 1 if workspace doesn\'t exists', async () => {
    jest.spyOn(project, 'workspace').mockResolvedValue(null);
    jest.spyOn(cmd, 'define').mockResolvedValue({ workspace: 'does-not-exists' });

    // Call
    await expect(cmd.run())
      .resolves.toBe(1);

    // Checks
    expect(transport.spinner.start).toHaveBeenCalledWith('Loading "does-not-exists" workspace');
    expect(project.workspace).toHaveBeenCalledWith('does-not-exists');
    expect(transport.spinner.fail).toHaveBeenCalledWith('Workspace "does-not-exists" not found');
  });

  it('should exit 1 if current workspace not found', async () => {
    jest.spyOn(project, 'workspace');
    jest.spyOn(project, 'currentWorkspace').mockResolvedValue(null);
    jest.spyOn(cmd, 'define').mockResolvedValue({ workspace: undefined });

    // Call
    await expect(cmd.run())
      .resolves.toBe(1);

    // Checks
    expect(transport.spinner.start).toHaveBeenCalledWith('Loading "." workspace');
    expect(project.workspace).not.toHaveBeenCalled();
    expect(project.currentWorkspace).toHaveBeenCalled();
    expect(transport.spinner.fail).toHaveBeenCalledWith('Workspace "." not found');
  });

  it('should print workspace basic info', async () => {
    const wks = new Workspace('./wks', { name: 'wks', version: '1.0.0' }, project);

    jest.spyOn(project, 'workspace').mockResolvedValue(wks);
    jest.spyOn(cmd, 'define').mockResolvedValue({ workspace: 'wks' });

    // Call
    await expect(cmd.run())
      .resolves.toBeUndefined();

    // Checks
    expect(transport.spinner.start).toHaveBeenCalledWith('Loading "wks" workspace');
    expect(project.workspace).toHaveBeenCalledWith('wks');
    expect(transport.spinner.stop).toHaveBeenCalled();
    expect(screen).toMatchSnapshot();
  });

  it('should use current workspace', async () => {
    const wks = new Workspace('./wks', { name: 'wks', version: '1.0.0' }, project);

    jest.spyOn(project, 'workspace');
    jest.spyOn(project, 'currentWorkspace').mockResolvedValue(wks);
    jest.spyOn(cmd, 'define').mockResolvedValue({ workspace: undefined });

    // Call
    await expect(cmd.run())
      .resolves.toBeUndefined();

    // Checks
    expect(transport.spinner.start).toHaveBeenCalledWith('Loading "." workspace');
    expect(project.workspace).not.toHaveBeenCalled();
    expect(project.currentWorkspace).toHaveBeenCalled();
  });

  it('should print workspace basic info with dependencies', async () => {
    jest.spyOn(cmd, 'define').mockResolvedValue({ workspace: 'wks' });
    jest.spyOn(project, 'workspace')
      .mockImplementation(async (name = 'wks') => new Workspace('./wks', { name, version: '1.0.0', dependencies: { depA: '1.0.0', depB: '1.0.0' } }, project));

    // Call
    await expect(cmd.run())
      .resolves.toBeUndefined();

    // Checks
    expect(transport.spinner.start).toHaveBeenCalledWith('Loading "wks" workspace');
    expect(project.workspace).toHaveBeenCalledWith('wks');
    expect(transport.spinner.stop).toHaveBeenCalled();
    expect(screen).toMatchSnapshot();
  });

  it('should print workspace basic info with dev-dependencies', async () => {
    jest.spyOn(cmd, 'define').mockResolvedValue({ workspace: 'wks' });
    jest.spyOn(project, 'workspace')
      .mockImplementation(async (name = 'wks') => new Workspace('./wks', { name, version: '1.0.0', devDependencies: { depA: '1.0.0', depB: '1.0.0' } }, project));

    // Call
    await expect(cmd.run())
      .resolves.toBeUndefined();

    // Checks
    expect(transport.spinner.start).toHaveBeenCalledWith('Loading "wks" workspace');
    expect(project.workspace).toHaveBeenCalledWith('wks');
    expect(transport.spinner.stop).toHaveBeenCalled();
    expect(screen).toMatchSnapshot();
  });
});
