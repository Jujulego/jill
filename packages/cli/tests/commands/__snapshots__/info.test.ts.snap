// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jill info should print workspace basic info 1`] = `
"Workspace [1mwks[22m:
[1mVersion:[22m   1.0.0
[1mDirectory:[22m wks

[1mDependencies:[22m
[90m   No dependencies for wks[39m
"
`;

exports[`jill info should print workspace basic info with dependencies 1`] = `
"Workspace [1mwks[22m:
[1mVersion:[22m   1.0.0
[1mDirectory:[22m wks

[1mDependencies:[22m
â”œâ”€ depA[90m@1.0.0[39m
â”‚  â”œâ”€ [3mdepA[90m@1.0.0[39m[23m
â”‚  â””â”€ [3mdepB[90m@1.0.0[39m[23m
â””â”€ [3mdepB[90m@1.0.0[39m[23m
"
`;

exports[`jill info should print workspace basic info with dev-dependencies 1`] = `
"Workspace [1mwks[22m:
[1mVersion:[22m   1.0.0
[1mDirectory:[22m wks

[1mDependencies:[22m
â”œâ”€ [34mdepA[90m@1.0.0[39m[34m (dev)[39m
â”‚  [34mâ”œâ”€ [39m[3m[34mdepA[90m@1.0.0[39m[34m (dev)[39m[23m
â”‚  [34mâ””â”€ [39m[3m[34mdepB[90m@1.0.0[39m[34m (dev)[39m[23m
â””â”€ [3m[34mdepB[90m@1.0.0[39m[34m (dev)[39m[23m
"
`;
